### PR TITLE
Parallelize S3 uploads

### PR DESCRIPTION
**Pull request by anishathalye**
With the Snacks dataset (848 MB, 6745 images) for example, this change brings the upload time down from 12 minutes to 1 minute.

### CODE REVIEW

**Comment by anishathalye** [comment]
Testing with a larger dataset like MNIST: one downside is that S3 doesn't seem to like us making so many requests in parallel (parameter is currently set to 100).

File: cleanlab_studio/cli/api_service.py [hunk 1]
```diff
@@ -19,6 +20,9 @@
 base_url = os.environ.get("CLEANLAB_API_BASE_URL", "https://api.cleanlab.ai/api/cli/v0")
 
 
+MAX_PARALLEL_UPLOADS = 100  # XXX choose this dynamically?
```
Comment (ryansingman) [inline cleanlab_studio/cli/api_service.py:23]: what would the optimal value for this depend on? are they things we can know before starting the upload?

I could see it making sense to do something like `min(<num_files> / <some_factor>, <max_allowed_parallel_uploads>)`

**Review by ryansingman — APPROVED** [review]
One comment but largely LGTM

**Comment by ryansingman** [comment]
> Testing with a larger dataset like MNIST: one downside is that S3 doesn't seem to like us making so many requests in parallel (parameter is currently set to 100).

That's strange. I'm seeing this [package](https://github.com/mishudark/s3-parallel-put/blob/master/s3-parallel-put) which uses 1024 processes to upload. It [apparently performs well](https://stackoverflow.com/questions/42235618/s3-how-to-upload-large-number-of-files). It is using `boto3` though, not post requests like we're doing

**Comment by anishathalye** [comment]
Some chance that the boto3 client uses a different endpoint, and AWS is applying different rate limiting rules for the two. ¯\_(ツ)_/¯ 

In any case, 30 seems to work. We can go with this and revisit later. Should probably implement backoff when getting 429 errors, retries on 500 errors, etc., which is missing from this PR.

**Comment by ryansingman** [comment]
KK, that sounds good to me 👍 . Should be good to merge once you add type checking to that nested function to make mypy happy

**Review by aayushchhabra1999 — APPROVED** [review]
LGTM! Maybe set MAX_PARALLEL_UPLOADS = 32 as a convention (Power of 2). I actually think static parallel uploaded are more than good enough till we learn we need to make that factor dynamic to make some dataset work.

**Comment by anishathalye** [comment]
Squashed and merged in e5c71a0.

**Pull request closed without merging**